src/adapter/src/coord/introspection.rs

```diff
@@ -80,6 +81,50 @@ pub fn auto_run_on_introspection<'a, 's>(
     }
 }
 
+/// Checks if we're currently running on the [`MZ_INTROSPECTION_CLUSTER`], and if so, do
+/// we depend on any objects that we're not allowed to query from the cluster.
+pub fn check_cluster_restrictions(
+    catalog: &impl SessionCatalog,
+    plan: &Plan,
+    depends_on: &Vec<GlobalId>,
+) -> Result<(), AdapterError> {
+    // We only impose restrictions if the current cluster is the introspection cluster.
+    let cluster = catalog.active_cluster();
+    if cluster != MZ_INTROSPECTION_CLUSTER.name {
+        return Ok(());
+    }
+
+    // Allows explain queries.
+    if let Plan::Explain(_) = plan {
```

Comment:
Minor concern: I'm not convinced this is helpful? We'd let them run an explain and then when they do a real query it'd error? Also fine leaving as is.

Reply:
The reason I opted to allow all explain queries was because I pictured a flow where we reject a query because it references non-system tables, and I wanted the user to be able to better understand why we rejected it, i.e. use `EXPLAIN`, without having to switch clusters. It's a minor detail and I'm totally fine with removing it if we think it'll cause more confusion because of the inconsistency